### ISSUE TITLE
Output parameters on steps template prevents graph in UI from showing following steps

### ISSUE THREAD

**Issue by daniebrill** [argoproj/argo-workflows#12247]
### Pre-requisites

- [X] I have double-checked my configuration
- [x] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

When calling a steps template with output parameters inside of a steps template, the steps after the steps template don't get rendered in the graph in the UI.
![grafik](https://github.com/argoproj/argo-workflows/assets/50454544/2a1b3ed1-52e2-4a54-8cc4-0b3e3fcda377)
In my example the last step (print) is missing.

### Version

v3.5.1

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: graph-bug
  namespace: argo
spec:
  entrypoint: main
  serviceAccountName: "executor"
  templates:
    - name: main
      steps:
        - - name: steps
            template: steps
        - - name: print
            template: whalesay
            arguments:
              parameters:
                - name: message
                  value: "{{steps.steps.outputs.parameters.test}}"
    - name: steps
      outputs:
        parameters:
          - name: test
            valueFrom:
              parameter: "{{steps.a.outputs.result}}"
      steps:
        - - name: a
            template: whalesay
            arguments:
              parameters:
                - name: message
                  value: '"first message"'
    - name: whalesay
      inputs:
        parameters:
          - name: message
      container:
        image: debian:bookworm-slim
        command: [ echo ]
        args: [ "{{inputs.parameters.message}}" ]
        resources:
          limits:
            memory: 32Mi
            cpu: 100m
```

### Logs from the workflow controller

```text
time="2023-11-22T16:36:31.506Z" level=info msg="Processing workflow" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.510Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.510Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.511Z" level=info msg="was unable to obtain node for , letting display name to be nodeName" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.511Z" level=info msg="Steps node graph-bug initialized Running" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.511Z" level=info msg="StepGroup node graph-bug-3097073270 initialized Running" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.511Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.511Z" level=info msg="Steps node graph-bug-1195036827 initialized Running" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.511Z" level=info msg="StepGroup node graph-bug-3894932087 initialized Running" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.511Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.512Z" level=info msg="Pod node graph-bug-3146999310 initialized Pending" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.535Z" level=info msg="Created pod: graph-bug[0].steps[0].a (graph-bug-whalesay-3146999310)" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.535Z" level=info msg="Workflow step group node graph-bug-3894932087 not yet completed" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.535Z" level=info msg="Workflow step group node graph-bug-3097073270 not yet completed" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.535Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.535Z" level=info msg=reconcileAgentPod namespace=argo workflow=graph-bug
time="2023-11-22T16:36:31.547Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1912710 workflow=graph-bug
time="2023-11-22T16:36:41.531Z" level=info msg="Processing workflow" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.531Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=graph-bug
time="2023-11-22T16:36:41.531Z" level=info msg="task-result changed" namespace=argo nodeID=graph-bug-3146999310 workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="node changed" namespace=argo new.message= new.phase=Succeeded new.progress=0/1 nodeID=graph-bug-3146999310 old.message= old.phase=Pending old.progress=0/1 workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="Step group node graph-bug-3894932087 successful" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="node graph-bug-3894932087 phase Running -> Succeeded" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="node graph-bug-3894932087 finished: 2023-11-22 16:36:41.532717393 +0000 UTC" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="Outbound nodes of graph-bug-3146999310 is [graph-bug-3146999310]" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="Outbound nodes of graph-bug-1195036827 is [graph-bug-3146999310]" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="node graph-bug-1195036827 phase Running -> Succeeded" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="node graph-bug-1195036827 finished: 2023-11-22 16:36:41.532764493 +0000 UTC" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="Step group node graph-bug-3097073270 successful" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="node graph-bug-3097073270 phase Running -> Succeeded" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="node graph-bug-3097073270 finished: 2023-11-22 16:36:41.532867992 +0000 UTC" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="StepGroup node graph-bug-1956048083 initialized Running" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=info msg="SG Outbound nodes of graph-bug-1195036827 are []" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.532Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.533Z" level=info msg="Pod node graph-bug-3400538558 initialized Pending" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.542Z" level=info msg="Created pod: graph-bug[1].print (graph-bug-whalesay-3400538558)" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.542Z" level=info msg="Workflow step group node graph-bug-1956048083 not yet completed" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.542Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.542Z" level=info msg=reconcileAgentPod namespace=argo workflow=graph-bug
time="2023-11-22T16:36:41.550Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=1912765 workflow=graph-bug
time="2023-11-22T16:36:41.560Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/graph-bug-whalesay-3146999310/labelPodCompleted
time="2023-11-22T16:36:51.545Z" level=info msg="Processing workflow" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.545Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=graph-bug
time="2023-11-22T16:36:51.545Z" level=info msg="task-result changed" namespace=argo nodeID=graph-bug-3146999310 workflow=graph-bug
time="2023-11-22T16:36:51.545Z" level=info msg="node changed" namespace=argo new.message= new.phase=Succeeded new.progress=0/1 nodeID=graph-bug-3400538558 old.message= old.phase=Pending old.progress=0/1 workflow=graph-bug
time="2023-11-22T16:36:51.545Z" level=info msg="SG Outbound nodes of graph-bug-1195036827 are []" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="Step group node graph-bug-1956048083 successful" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="node graph-bug-1956048083 phase Running -> Succeeded" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="node graph-bug-1956048083 finished: 2023-11-22 16:36:51.546123731 +0000 UTC" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="Outbound nodes of graph-bug-3400538558 is [graph-bug-3400538558]" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="Outbound nodes of graph-bug is [graph-bug-3400538558]" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="node graph-bug phase Running -> Succeeded" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="node graph-bug finished: 2023-11-22 16:36:51.546319256 +0000 UTC" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg=reconcileAgentPod namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="Updated phase Running -> Succeeded" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.546Z" level=info msg="Marking workflow completed" namespace=argo workflow=graph-bug
time="2023-11-22T16:36:51.552Z" level=info msg="cleaning up pod" action=deletePod key=argo/graph-bug-1340600742-agent/deletePod
time="2023-11-22T16:36:51.556Z" level=info msg="Workflow update successful" namespace=argo phase=Succeeded resourceVersion=1912820 workflow=graph-bug
time="2023-11-22T16:36:51.569Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/graph-bug-whalesay-3400538558/labelPodCompleted
```

### Logs from in your workflow's wait container

```text
time="2023-11-22T16:36:34.154Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
time="2023-11-22T16:36:34.154Z" level=info msg="Executor initialized" deadline="0001-01-01 00:00:00 +0000 UTC" includeScriptOutput=true namespace=argo podName=graph-bug-whalesay-3146999310 template="{\"name\":\"whalesay\",\"inputs\":{\"parameters\":[{\"name\":\"message\",\"value\":\"\\\"first message\\\"\"}]},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"debian:bookworm-slim\",\"command\":[\"echo\"],\"args\":[\"\\\"first message\\\"\"],\"resources\":{\"limits\":{\"cpu\":\"100m\",\"memory\":\"32Mi\"}}}}" version="&Version{Version:v3.5.1,BuildDate:2023-11-03T18:37:58Z,GitCommit:877c5523066e17687856fe3484c9b2d398e986f5,GitTag:v3.5.1,GitTreeState:clean,GoVersion:go1.21.3,Compiler:gc,Platform:linux/amd64,}"
time="2023-11-22T16:36:34.155Z" level=info msg="Starting deadline monitor"
time="2023-11-22T16:36:36.155Z" level=info msg="Main container completed" error="<nil>"
time="2023-11-22T16:36:36.155Z" level=info msg="Capturing script output"
time="2023-11-22T16:36:36.155Z" level=info msg="No output parameters"
time="2023-11-22T16:36:36.155Z" level=info msg="No output artifacts"
time="2023-11-22T16:36:36.167Z" level=info msg="Alloc=7110 TotalAlloc=13070 Sys=18789 NumGC=4 Goroutines=8"
time="2023-11-22T16:36:36.167Z" level=info msg="stopping progress monitor (context done)" error="context canceled"
time="2023-11-22T16:36:36.167Z" level=info msg="Deadline monitor stopped"
time="2023-11-22T16:36:44.222Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
time="2023-11-22T16:36:44.222Z" level=info msg="Executor initialized" deadline="0001-01-01 00:00:00 +0000 UTC" includeScriptOutput=false namespace=argo podName=graph-bug-whalesay-3400538558 template="{\"name\":\"whalesay\",\"inputs\":{\"parameters\":[{\"name\":\"message\",\"value\":\"\\\"first message\\\"\"}]},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"debian:bookworm-slim\",\"command\":[\"echo\"],\"args\":[\"\\\"first message\\\"\"],\"resources\":{\"limits\":{\"cpu\":\"100m\",\"memory\":\"32Mi\"}}}}" version="&Version{Version:v3.5.1,BuildDate:2023-11-03T18:37:58Z,GitCommit:877c5523066e17687856fe3484c9b2d398e986f5,GitTag:v3.5.1,GitTreeState:clean,GoVersion:go1.21.3,Compiler:gc,Platform:linux/amd64,}"
time="2023-11-22T16:36:44.223Z" level=info msg="Starting deadline monitor"
time="2023-11-22T16:36:45.225Z" level=info msg="Main container completed" error="<nil>"
time="2023-11-22T16:36:45.225Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2023-11-22T16:36:45.225Z" level=info msg="No output parameters"
time="2023-11-22T16:36:45.225Z" level=info msg="No output artifacts"
time="2023-11-22T16:36:45.225Z" level=info msg="Alloc=6942 TotalAlloc=12776 Sys=19045 NumGC=4 Goroutines=7"
time="2023-11-22T16:36:45.225Z" level=info msg="stopping progress monitor (context done)" error="context canceled"
time="2023-11-22T16:36:45.225Z" level=info msg="Deadline monitor stopped"
```

**Comment by Joibel** [comment]
I am pretty sure this is fixed by #12203 which will be in the 3.5.2. If you can re-test once that is released (or test on 3.5.0 which doesn't have the bug) we can confirm

**Comment by daniebrill** [comment]
Hi @Joibel, thanks for your help. The problem is solved by using v3.5.0.

**Comment by agilgur5** [comment]
For posterity, marking as duplicate of #12165
